flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JDBCTableSourceSinkFactory.java

```diff
@@ -0,0 +1,228 @@
+		final Optional<Long> partitionLower = descriptorProperties.getOptionalLong(CONNECTOR_SCAN_LOWER_BOUND);
+		final Optional<Long> partitionUpper = descriptorProperties.getOptionalLong(CONNECTOR_SCAN_UPPER_BOUND);
+		final Optional<Long> numPartitions = descriptorProperties.getOptionalLong(CONNECTOR_SCAN_NUM_PARTITIONS);
+		final JDBCScanOptions scanOptions;
+		if (partitionColumnName.isPresent()) {
+			Preconditions.checkArgument(
+				numPartitions.isPresent() && partitionLower.isPresent() && partitionUpper.isPresent(),
+				"Either all or none of the properties of scan options should be provided");
+			scanOptions = JDBCScanOptions.builder()
+				.setPartitionColumnName(partitionColumnName.get())
+				.setPartitionLowerBound(partitionLower.get())
+				.setPartitionUpperBound(partitionUpper.get())
+				.setNumPartitions(numPartitions.get())
+				.build();
+		} else {
+			scanOptions = null;
+		}
+
+		final Optional<Long> lookupCacheMaxSize =
+			descriptorProperties.getOptionalLong(CONNECTOR_LOOKUP_CACHE_MAX_ENTRIES);
+		final Optional<TimeLength> lookupCacheExpireMs =
+			descriptorProperties.getOptionalTimeLength(CONNECTOR_LOOKUP_CACHE_TTL);
+		final Optional<Integer> lookupMaxRetryTimes =
+			descriptorProperties.getOptionalInt(CONNECTOR_LOOKUP_MAX_RETRIES);
+		JDBCLookupOptions.Builder lookupOptionsBuilder = JDBCLookupOptions.builder();
+		if (lookupCacheMaxSize.isPresent()) {
+			lookupOptionsBuilder = lookupOptionsBuilder.setCacheMaxSize(lookupCacheMaxSize.get());
+		}
+		if (lookupCacheExpireMs.isPresent()) {
+			lookupOptionsBuilder = lookupOptionsBuilder.setCacheExpireMs(lookupCacheExpireMs.get().getMilliseconds());
+		}
+		if (lookupMaxRetryTimes.isPresent()) {
+			lookupOptionsBuilder = lookupOptionsBuilder.setMaxRetryTimes(lookupMaxRetryTimes.get());
+		}
+		final JDBCLookupOptions lookupOptions = lookupOptionsBuilder.build();
+
+		final TableSchema schema = descriptorProperties.getTableSchema(SCHEMA);
+		return JDBCTableSource.builder()
+				.setOptions(options)
+				.setScanOptions(scanOptions)
+				.setLookupOptions(lookupOptions)
+				.setSchema(schema)
+				.build();
+	}
+
+	@Override
+	public StreamTableSink<Tuple2<Boolean, Row>> createStreamTableSink(Map<String, String> properties) {
+		final DescriptorProperties descriptorProperties = getValidatedProperties(properties);
+
+		JDBCOptions options = getJDBCOptions(descriptorProperties);
+		JDBCUpsertTableSink.Builder builder = JDBCUpsertTableSink.builder()
+			.setOptions(options)
+			.setTableSchema(descriptorProperties.getTableSchema(SCHEMA));
+
+		Optional<Integer> flushMaxSize = descriptorProperties.getOptionalInt(CONNECTOR_WRITE_FLUSH_MAX_ENTRIES);
+		Optional<TimeLength> flushIntervalMs = descriptorProperties.getOptionalTimeLength(CONNECTOR_WRITE_FLUSH_INTERVAL);
+		Optional<Integer> maxRetryTimes = descriptorProperties.getOptionalInt(CONNECTOR_WRITE_MAX_RETRIES);
+		if (flushMaxSize.isPresent()) {
+			builder = builder.setFlushMaxSize(flushMaxSize.get());
```

Comment:
You don't need to reassign to builder var, builder return value just for `builder.set1.set2.set3`.